proposals/003-authentication-api.md

```diff
@@ -0,0 +1,516 @@
+     */
+    void clientAuthenticationFailure(LoginException exception);
+    
+}
+```
+
+It's worth noting that `LoginException` has a number of subclasses in `javax.security.auth.login`.
+
+This API allows SASL-terminating plugin to announce its authentication outcomes to later filters in the filter chain which have implemented `ClientSubjectAware`.
+Note that client authentication outcomes only propagate towards the server, not back towards the client. 
+So a `ClientSubjectAware` before a SASL-terminating plugin will not receive the announcement.
+**TODO justify this**
+
+
+### Example: A SASL Terminating `Filter`
+
+This shows how a SASL terminating `Filter` would use the new methods on `FilterContext` to inform other filters, such as `ExampleIdentityConsumingFilter` above, about the client.
+
+The proxy config would look like this:
+
+```yaml
+virtualClusters:
+  - name: my-cluster
+    filters:
+      - my_authn_filter
+      - my_authz_filter
+filterDefinitions:
+   - name: my_authn_filter
+     type: ExampleSaslTerminatingFilter
+     config:
+       jaasConfigFile: client-jaas.login.config
+       jaasContextName: client_auth_stack
+   - name: my_authz_filter
+     type: ExampleIdentityConsumingFilter
+     config: ...
+```
+
+where the `client-jaaas.login.config` file might look like this:
+
+```
+client_auth_stack {
+  org.apache.kafka.common.security.plain.PlainLoginModule required
+     user_alice="pa55word"
+     user_bob="changeit"
+     ;
+};
+```
+
+The implementation would look something like this:
+
+```java
+public class ExampleSaslTerminatingFilter implements SaslAuthenticateRequestFilter, ClientSubjectAware {
+
+    private Subject subject;
+
+    @Override
+    public void onClientAuthentication(Subject clientSubject, FilterContext context) {
+        // the clientSubject will have an X500Principal iff the client used mTLS.
+        this.subject = clientSubject;
+    }
+
+    @Override
+    public CompletionStage<RequestFilterResult> onSaslAuthenticateRequest(short apiVersion,
+                                                                   RequestHeaderData header,
+                                                                   SaslAuthenticateRequestData request,
+                                                                   FilterContext context) {
+        Configuration config = new ConfigFile(URI.create("client-jaas.login.config"));
+
+        // TODO the callback handler to use depends on the context in the jaas config
+        //   so how do we know which handler to instantiate?
+        //   Kafka's SaslChannelBuilder basically does its own parsing of the jaas configuration to figure it out.
+        CallbackHandler callbackHandler = new PlainServerCallbackHandler();
+
+        try {
+            LoginContext loginContext = new LoginContext("client_auth_stack", subject, callbackHandler, config);
+            loginContext.login();
```

Comment:
aside: I _suspect_ this call may block, depending on the JAAS configuration.   I know this is just example code, but it probably ought to use its own TPE.  Adding a comment will be sufficient for the proposal.  If the code ever made it is into documentation, we'd want a async approach..

Reply:
You're correct that we can't rule out Callbackhandlers from blocking. The method return type is `CompletionStage<RequestFilterResult>` for that reason, but I can add a comment.